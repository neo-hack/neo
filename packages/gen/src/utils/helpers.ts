import { Workflow } from '../interface'
import { hooks } from '../utils/hooks'

import Observable from 'zen-observable'
import type Listr from 'listr'
import { LIFE_CYCLES } from 'src/constants'

/**
 * @description convert schema to listr
 */
export const toListr = (schema: Workflow) => {
  const tasks: Listr.ListrTask[] = Object.keys(schema.jobs || {}).map((key) => {
    const taskName = schema.jobs![key].name || key
    return {
      title: taskName,
      task: () => {
        let ob: ZenObservable.SubscriptionObserver<any>
        const OB = new Observable<any>((observer) => {
          ob = observer
        })
        hooks.addHooks({
          [LIFE_CYCLES.STEP]: (args: { step: string }) => {
            ob.next(args.step)
          },
          [taskName]: () => {
            ob.complete()
          },
        })
        return OB as any
      },
    }
  })
  return tasks
}
