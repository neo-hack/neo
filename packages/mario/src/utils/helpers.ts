import { Workflow } from '../interface'
import { hooks } from './hooks'

import Observable from 'zen-observable'
import type Listr from 'listr'
import { LIFE_CYCLES } from 'src/constants'

/**
 * @description convert schema to listr
 */
export const toListr = (schema: Workflow) => {
  const tasks: Listr.ListrTask[] = Object.keys(schema.jobs || {}).map((key) => {
    const taskName = schema.jobs![key].name || key
    return {
      title: taskName,
      task: () => {
        let ob: ZenObservable.SubscriptionObserver<any>
        const OB = new Observable<any>((observer) => {
          ob = observer
        })
        hooks.addHooks({
          [LIFE_CYCLES.STEP]: (args: { step: string }) => {
            ob.next(args.step)
          },
          [taskName]: (args: { error: Error }) => {
            if (!args.error) {
              ob.complete()
              return
            }
            ob.error(args.error)
          },
        })
        return OB as any
      },
    }
  })
  return tasks
}
