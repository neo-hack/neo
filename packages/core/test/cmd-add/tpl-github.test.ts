import { execNeo, readLockFile, clearLockFile, storeDir } from '../helpers'

afterEach(() => {
  clearLockFile()
})

it('load template from github should work', async () => {
  const url = 'https://github.com/neo-hack/actions-template'
  await execNeo(['add', url, '--store-dir', storeDir])
  expect(readLockFile()).toMatchSnapshot()
})
