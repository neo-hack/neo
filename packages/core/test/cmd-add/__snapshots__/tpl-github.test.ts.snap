// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`load template from github should work 1`] = `
"templates:
  github.com/neo-hack/actions-template/dd4b6c410813dd331f46a964dd0d34cd7d59e4a7:
    name: '@aiou/actions-template'
    version: 0.2.0
    resolvedVia: git-repository
    id: >-
      github.com/neo-hack/actions-template/dd4b6c410813dd331f46a964dd0d34cd7d59e4a7
    pref: https://github.com/neo-hack/actions-template
"
`;
